groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -218,16 +219,243 @@ extern const Optional_OptNoSuchType optNoSuchType;
     // functions that are constrained using a function argument.
 
 #ifdef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+
+#define BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(U, V)                           \
+    || std::is_constructible<U, V>::value
+#define BSLSTL_OPTIONAL_AND_IS_CONSTRUCTIBLE_V(U, V)                          \
+    &&std::is_constructible<U, V>::value
+#define BSLSTL_OPTIONAL_AND_IS_ASSIGNABLE_V(U, V)                             \
+    &&std::is_assignable<U, V>::value
 #define BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE                             \
     std::is_trivially_destructible
 
 #else
+
+#define BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(U, V)
+#define BSLSTL_OPTIONAL_AND_IS_CONSTRUCTIBLE_V(U, V)
+#define BSLSTL_OPTIONAL_AND_IS_ASSIGNABLE_V(U, V)
 #define BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE bsl::is_trivially_copyable
     // C++03 does not provide a trivially destructible trait.  Instead we use
     // 'bsl::is_trivially_copyable' which implies the type is also trivially
     // destructible.
 #endif  // BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
 
+// Type traits to assist in choosing the correct assignment and construction
+// overload.  If the 'value_type' converts  or assigns from an
+// 'optional<other_type>', then the overload passing the function parameter to
+// the 'value_type' is preferred.  As in 'std' implementation, if the
+// 'value_type' converts or assigns from any value category, we consider it
+// convertible/assignable from optional.
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromStdOptional
```

Comment:
Can we avoid repeating these checks for the two optional types?

```c++
struct Optional_ConvertsFromOptional_Impl
: bsl::integral_constant<
    bsl::is_convertible<const bsl::optional<ANY_TYPE>&, TYPE>::value 
    /* ... */> {
};

#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
template <class TYPE, class ANY_TYPE>
struct Optional_ConvertsFromStdOptional
: Optional_ConvertsFromOptional_Impl<TYPE, ANY_TYPE, std::optional> {
};

#else
template <class TYPE, class ANY_TYPE>
struct Optional_ConvertsFromStdOptional : bsl::integral_constant<bool, false> {
};
#endif  // BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY

template <class TYPE, class ANY_TYPE>
struct Optional_ConvertsFromOptional
: bsl::integral_constant<
      bool,
      Optional_ConvertsFromOptional_Impl<TYPE, ANY_TYPE, bsl::optional>::value
          || Optional_ConvertsFromStdOptional<TYPE, ANY_TYPE>::value> {
};
```

Reply:
I've done something along those lines.

Reply:
Looks good